pkg/validators/quota.go

```diff
@@ -86,149 +88,150 @@ func validateResourceRequirements(rs []ResourceRequirement, up *usageProvider) (
 		groups[k] = append(groups[k], r)
 	}
 
-	for k, g := range groups {
-		ls, err := serviceLimits(k.Consumer, k.Service)
-		if err != nil {
-			return qe, err
-		}
-		qse, err := validateServiceLimits(g, ls, up)
-		if err != nil {
-			return qe, err
-		}
-		qe = append(qe, qse...)
+	// Process all groups in parallel
+	type chs struct {
+		qe  []QuotaError
+		err error
 	}
-
-	return qe, nil
-}
-
-func validateServiceLimits(rs []ResourceRequirement, ls []*sub.ConsumerQuotaMetric, up *usageProvider) ([]QuotaError, error) {
-	// Group by Metric and Aggregation
-	type gk struct {
-		Metric      string
-		Aggregation string
+	ch := make(chan chs)
+	for k, g := range groups { // Spawn
+		go func(k gk, g []ResourceRequirement) {
+			qe, err := validateServiceRequirements(k.Consumer, k.Service, g, up)
+			ch <- chs{qe, err}
+		}(k, g)
 	}
-	groups := map[gk][]ResourceRequirement{}
-	for _, r := range rs {
-		k := gk{r.Metric, r.Aggregation}
-		groups[k] = append(groups[k], r)
+	errs := config.Errors{}
+	qerrs := []QuotaError{}
+	for range groups { // Gather
+		s := <-ch
+		qerrs = append(qerrs, s.qe...)
+		errs.Add(s.err)
 	}
+	return qerrs, errs.OrNil()
+}
 
-	qe := []QuotaError{}
-	for k, g := range groups {
-		agg, err := aggregation(k.Aggregation)
-		if err != nil {
-			return qe, err
-		}
+// Validate requirements for a single consume & service pair.
+func validateServiceRequirements(consumer string, service string, rs []ResourceRequirement, up *usageProvider) ([]QuotaError, error) {
+	qms, err := queryMetrics(consumer, service)
+	if err != nil {
+		return nil, err
+	}
 
-		// select limits for the metric
-		ml := []*sub.ConsumerQuotaLimit{}
-		for _, l := range ls {
-			if l.Metric == k.Metric {
-				ml = append(ml, l.ConsumerQuotaLimits...)
-			}
-		}
-		if len(ml) == 0 {
-			return qe, fmt.Errorf("limits for metric %q were not found", k.Metric)
-		}
+	errs := config.Errors{}
+	reqToBuckets, err := gatherBucketsRequirements(rs, qms)
+	errs.Add(err)
+	qerrs := []QuotaError{}
+	for _, br := range reqToBuckets {
+		qerrs = append(qerrs, validateBucket(br, up)...)
+	}
+	return qerrs, errs.OrNil()
+}
 
-		for _, limit := range ml {
-			qle := validateLimit(g, limit, up, agg)
-			qe = append(qe, qle...)
+// Find a bucket in the ConsumerQuotaLimit that matches the ResourceRequirement.
+func findBucket(r ResourceRequirement, ql *sub.ConsumerQuotaLimit) (*sub.QuotaBucket, error) {
+	// Iterate buckets in order from most to less specific
+	for i := len(ql.QuotaBuckets) - 1; i >= 0; i-- {
+		if r.InBucket(ql.QuotaBuckets[i]) {
+			return ql.QuotaBuckets[i], nil
```

Comment:
It seems like this is only finding a single bucket for a given quota limit (greedy return). From our discussion I thought it would fine all possible buckets. Can you please clarify. If we are only taking a single bucket and not all buckets then what is the criteria for the single bucket? To me it looks like you are just running through the list backwards. Is there only guaranteed to be a single match? Is the list sorted from least to most specific? Consider renaming to `findBestMatchBucket`.

Reply:
> From our discussion I thought it would fine all possible buckets. 

We find multiple buckets, but exactly one per `ConsumerQuotaLimit`, the `ConsumerQuotaLimit` defines the "scope" of quota, e.g. regional or zonal.
E.g. for `N2_CPUs`  we will check two buckets, one from regional-scope and another from zonal-scope.

Reply:
>  If we are only taking a single bucket and not all buckets then what is the criteria for the single bucket? To me it looks like you are just running through the list backwards. Is the list sorted from least to most specific?

There is nothing in the function that leads me to believe we are returning the most specific match, it just looks like we are returning any match. Maybe I am missing something simple. Is the list known to be sorted with some special properties?